source/known-hook-events.ts

```diff
@@ -1,8 +1,81 @@
-export default [
+import {IncomingMessage} from 'http';
+import {Options, CancelableRequest} from './utils/types';
+import {HTTPError} from './errors';
+
+export type InitHook = (options: Options) => void;
+// TODO: The `Error` type should confirm to any possible extended error type that can be thrown. See https://github.com/sindresorhus/got#hooksbeforeerror
+export type BeforeErrorHook = (error: Error) => Error | void | Promise<Error> | Promise<void>;
+export type BeforeRequestHook = (options: Options) => void | Promise<void>
+export type BeforeRedirectHook = (options: Options) => void | Promise<void>
+export type BeforeRetryHook = (options: Options, error: HTTPError, retryCount: number) => void | Promise<void>
+export type AfterResponseHook = (response: IncomingMessage, retryWithMergedOptions: (options: Options) => IncomingMessage | CancelableRequest<IncomingMessage>) => IncomingMessage | CancelableRequest<IncomingMessage>
+
+export type HookType = BeforeErrorHook | InitHook | BeforeRequestHook | BeforeRedirectHook | BeforeRetryHook | AfterResponseHook;
+
+/**
+ * Hooks allow modifications during the request lifecycle. Hook functions may be async and are run serially.
+ */
```

Comment:
@sindresorhus Do we need to double the documentation?

Reply:
https://github.com/sindresorhus/got/pull/714#discussion_r251441415

Reply:
Agreed on that thing. Sorry I didn't shape my question correctly: I mean, generally, do we need to redocument *everything* in the code, isn't the README just enough?

Reply:
I think the issue with using the README is that you quickly end up with out of date documentation.

For example in the documentation for [`beforeError`](https://github.com/sindresorhus/got/#hooksbeforeerror) hook, the example uses `onError` as the property name—this could be a typo or a change that didn't get updated, but nevertheless it makes it more difficult.

If we redocument everything, we can later generate the documentation.

Reply:
> If we redocument everything, we can later generate the documentation.

Looks promising! Let's do that way then :)